src/common/test-webserver.c

```diff
@@ -1168,5 +1175,66 @@ main (int argc,
   g_test_add ("/web-server/protocol-header/ignore", TestFixture, &test_case_proto_header_ignore,
               setup, test_webserver_with_headers, teardown);
 
+
+  /* Header is enabled, but not passed. */
+  static const TestCase test_case_forwarded_for_empty = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .expected_remote = "127.0.0.1",
+  };
+  g_test_add ("/web-server/forwarded-for/empty", TestFixture, &test_case_forwarded_for_empty,
+              setup, test_webserver_with_headers, teardown);
+
+  /* Header enabled, and passed an IPv4 address */
+  static const TestCase test_case_forwarded_for_v4 = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .extra_headers = "X-Forwarded-For: 1.2.3.4\r\n",
+    .expected_remote = "1.2.3.4",
+  };
+  g_test_add ("/web-server/forwarded-for/v4", TestFixture, &test_case_forwarded_for_v4,
+              setup, test_webserver_with_headers, teardown);
+
+  /* Header enabled, and passed an IPv6 address */
+  static const TestCase test_case_forwarded_for_v6 = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .extra_headers = "X-Forwarded-For: 2001::1\r\n",
+    .expected_remote = "2001::1",
+  };
+  g_test_add ("/web-server/forwarded-for/v6", TestFixture, &test_case_forwarded_for_v6,
+              setup, test_webserver_with_headers, teardown);
+
+  /* Header enabled, and passed 'unknown' */
+  static const TestCase test_case_forwarded_for_unknown = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .extra_headers = "X-Forwarded-For: unknown\r\n",
+    .expected_remote = "unknown",
+  };
+  g_test_add ("/web-server/forwarded-for/unknown", TestFixture, &test_case_forwarded_for_unknown,
+              setup, test_webserver_with_headers, teardown);
+
+  /* Header enabled, and passed multiple IPs */
+  static const TestCase test_case_forwarded_for_multiple = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .extra_headers = "X-Forwarded-For: 6.6.6.6 2.2.2.2 1.2.3.4\r\n",
+    .expected_remote = "1.2.3.4",
+  };
+  g_test_add ("/web-server/forwarded-for/multiple", TestFixture, &test_case_forwarded_for_multiple,
+              setup, test_webserver_with_headers, teardown);
+
+  /* Header enabled, and passed multiple IPs, and junk */
+  static const TestCase test_case_forwarded_for_junk = {
+    .forwarded_for_header = "X-Forwarded-For",
+    .extra_headers = "X-Forwarded-For: !@{}\"#%^&*()<>?`~\\|'$\t $whatever;   ;; ,,,  1.2.3.4\r\n",
```

Comment:
Would also be good to test the other edge cases: empty string and a lone space.

Reply:
OK.

Reply:
This actually shook out some bad behaviour: the reported remote host was `""` which is kinda useless.  I've added a check to ignore the header if the value is empty, so in that case we'll get the fallback to the IP of the proxy itself.